### PR TITLE
Add support for transpiling per-file jsx pragmas

### LINKED ISSUE #21114: Not to remove unused  import module ?
**TypeScript Version:** 2.4.1

**Code**
```
// tsconfig.json
{
...
"jsx": "preserve",
...
}
```
```
// my preact Component
import {h, Component} from 'preact'
class C extends Component<any, any> {
    render(){
    // ...
    }
}
```
```
// after compiled
import { Component } from 'preact'
```
So it will throw an error, such as "h is not defined"

**Expected behavior:**
```
// after compiled
import {h, Component } from 'preact'
```
**Actual behavior:**
```
// after compiled
import {Component} from 'preact' 
```
**Related:**

### PR DESCRIPTION
**Pull request by weswigham**
This adds support for per-file jsx factory pragmas (similar to babel):
```ts
/** @jsx dom */
import { dom } from "./renderer"
<h></h>
```
becomes
```js
var renderer_1 = require("./renderer");
renderer_1.dom("h", null);
```
We still expect the global `jsx` compiler option to be set and the file extension to be tsx or jsx; but now you can mix multiple jsx frameworks in one project (just not more than one per file). This only affects emit and not typechecking - multiple different frameworks' types attempting to define, eg, `JSX.Element` can still result in type errors if those definitions conflict.

Fixes #21114, probably?
ref #18131

### CODE REVIEW

**Comment by weswigham** [comment]
@uniqueiniquity A similar error is now present when a fragment is used alongside an inline jsx pragma. 🌞

**Comment by weswigham** [comment]
@mhegazy Do you wanna give this a brief look-over before it's merged?

File: tests/cases/conformance/jsx/inline/inlineJsxFactoryWithFragmentIsError.tsx [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+declare global {
+    namespace JSX {
+        interface IntrinsicElements {
+            [e: string]: any;
+        }
+    }
+}
+export function dom(): void;
+// @filename: index.tsx
+/** @jsx dom */
```
Comment (mhegazy) [inline tests/cases/conformance/jsx/inline/inlineJsxFactoryWithFragmentIsError.tsx]: add a test with `** @jsx React.createElement */`

File: src/compiler/parser.ts [hunk 2]
```diff
@@ -6089,6 +6089,7 @@ namespace ts {
             const referencedFiles: FileReference[] = [];
             const typeReferenceDirectives: FileReference[] = [];
             const amdDependencies: { path: string; name: string }[] = [];
+            let pragmas: { pragma: string, value: string | undefined }[] = [];
```
Comment (mhegazy) [inline src/compiler/parser.ts]: if we gonna add a new concept, then we should consolidate the other ones under pargma.. and we should allow `///` as well as multi-line comments.

I would also have helper functions for getting all of these like amdDependecies, amdModuleNAme, checkJs, etc..

File: src/compiler/checker.ts [hunk 3]
```diff
@@ -765,7 +765,22 @@ namespace ts {
...
 
-        function getJsxNamespace(): __String {
+        function getJsxNamespace(location: Node | undefined): __String {
+            if (location) {
+                const file = getSourceFileOfNode(location);
+                if (file) {
+                    if (file.localJsxNamespace) {
+                        return file.localJsxNamespace;
+                    }
+                    const jsxPragma = file.pragmas.get("jsx");
```
Comment (mhegazy) [inline src/compiler/checker.ts]: why is it `jsx` and not `jsxNamespace` like the compiler option?

File: tests/cases/conformance/jsx/inline/inlineJsxFactoryWithFragmentIsError.tsx [hunk 4]
```diff
@@ -0,0 +1,14 @@
...
+// @filename: renderer.d.ts
+declare global {
+    namespace JSX {
+        interface IntrinsicElements {
+            [e: string]: any;
+        }
+    }
+}
+export function dom(): void;
+// @filename: index.tsx
```
Comment (mhegazy) [inline tests/cases/conformance/jsx/inline/inlineJsxFactoryWithFragmentIsError.tsx]: also add a test with two files, and `@jsxNamespace` defined, and one of them with the new pragma.

File: src/compiler/checker.ts (hunk 3 above)
Comment (weswigham) [reply src/compiler/checker.ts]: Babel supports `jsx` and I figured we'd want to support the same thing - we can alias `jsxNamespace`, if you'd like?

File: src/compiler/parser.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/parser.ts]: Sure! I wanted to consolidate them, anyway, just didn't think this PR was appropriate.

Comment (sheetalkamat) [reply src/compiler/parser.ts]: We need to ensure this works well when doing incremental parsing as well and need test for the same.

**Comment by weswigham** [comment]
@mhegazy @sheetalkamat I've unified all pragma-parsing infrastructure such that new pragmas can be added just by editing an object literal, alternate syntaxes enabled, etc - also this PR is much larger now.

@mhegazy Which old/new pragmas did you want to allow under which syntaxes?

@sheetalkamat I think now that I've unified the implementations, incremental is handled correctly - `preProcess.ts` is the implementation of re-scanning this data for incremental purposes, right? Now it handles the new pragmas the same way as the old ones (and the same way as the parser, for that matter).

File: src/compiler/checker.ts [hunk 5]
```diff
@@ -765,7 +765,23 @@ namespace ts {
...
+                    if (jsxPragma) {
+                        const chosenpragma = isArray(jsxPragma) ? jsxPragma[0] : jsxPragma;
```
Comment (mhegazy) [inline src/compiler/checker.ts]: why not set this one in `processPragmasIntoFields` like the other ones, just for consistency.

Comment (weswigham) [reply src/compiler/checker.ts]: It pretty much maps to `localJsxFactory`, it's calculation is just deferred.

Comment (weswigham) [reply src/compiler/checker.ts]: I could move the calculation from the checker into the parser, if you'd like.

**Pull request merged**